### PR TITLE
first, last, skip, take, while, toObject

### PR DESCRIPTION
**Pull request by AntonStaravoitau**
first, last, skip, take, while, toObject

### CODE REVIEW

File: answers.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
+Array.prototype.first = function() {
+    return this[0];
+}
```
Comment (Shalimov) [inline answers.js]: As a suggestion: use semicolon to close expressions
it relates to all expressions below as well

File: answers.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+Array.prototype.skip = function(start) {
+    return this.slice(start);
+}
+Array.prototype.take = function(end) {
+    return this.slice(0, end);
+}
+Array.prototype.while = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
+        return null;
```
Comment (Shalimov) [inline answers.js]: Good job to check passing params, but
here are two possible ways how to handle it
first off - throw the exception
second one - return object itself
not so good to use logger inside (cuz it's dependency)

File: answers.js [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    var newArray = [];
+    for (let i = 0; i < this.length; i++) {
```
Comment (Shalimov) [inline answers.js]: if u use `let` why don't u use `let declaration` everywhere?

File: answers.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        if (predicate(this[i], i, this)) {
+            newArray.push(this[i]);
+        }
+    }
+    return newArray;
+}
```
Comment (Shalimov) [inline answers.js]: Implementation is wrong, cuz `while` should cut off the sequence if predicate `return false`
right now it works as filter function which is applied over array

File: answers.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        return null;
+    }
+    var newObject = {};
+    var tmpObject, tmpKey, tmpValue;
+    for (let i = 0; i < this.length; i++) {
+        tmpObject = predicate(this[i], i, this);
+        newObject[tmpObject.key] = tmpObject.value;
+    }
+    return newObject;
+}
```
Comment (Shalimov) [inline answers.js]: As agreed, skip review of `toObject` until task accomplishment

File: answers.js [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    let newArray = [];
+    for (let i = 0; i < this.length; i++) {
+        if (predicate(this[i], i, this)) {
+            newArray.push(this[i]);
+        } else {
+            break;
+        }
+    }
+    return newArray;
+};
```
Comment (Shalimov) [inline answers.js]: As an improvement:
```javascript

if(typeof predicate !== 'function') {
 return this; 
//or throw new Error('Predicate should be a function');
}

let newArray = [];

for(let i = 0, length = this.length; i < length; i += 1) {

 if(predicate(this[i], i, this) === false) {
  return newArray;
 }

 newArray.push(this[i]);
}

```

File: answers.js [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+Array.prototype.filter = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
+        return this;
+    }
```
Comment (Shalimov) [inline answers.js]: As I mentioned, `console.log` is dependency and u should avoid deps as much as possible

File: answers.js [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+Array.prototype.toObject = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
+        return this;
+    }
+    let newObject = {};
+    let tmpObject, tmpKey, tmpValue;
+    for (let i = 0; i < this.length; i++) {
+        tmpObject = predicate(this[i], i, this);
```
Comment (Shalimov) [inline answers.js]: why do you declare variable above and not here?

File: answers.js [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+    return newArray;
+};
+
+Array.prototype.toObject = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
+        return this;
+    }
+    let newObject = {};
+    let tmpObject, tmpKey, tmpValue;
```
Comment (Shalimov) [inline answers.js]: tmpKey, tmpValue are defined but not used

File: answers.js [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    for (let i = 0; i < this.length; i++) {
+        tmpObject = predicate(this[i], i, this);
+        newObject[tmpObject.key] = tmpObject.value;
+    }
+    return newObject;
+};
+
+Array.prototype.fold = function(predicate, initialValue) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
```
Comment (Shalimov) [inline answers.js]: @AntonStaravoitau Check comments above

File: answers.js [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+        return this;
+    }
+    let previousValue, currentValue;
+    if (typeof initialValue === 'number' && initialValue === initialValue) {
+        previousValue = this[0] + initialValue;
```
Comment (Shalimov) [inline answers.js]: This is wrong!
What for u do it?

File: answers.js [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+};
+
+Array.prototype.transform = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
+        return this;
+    }
+    let newArray = [];
+    for (let i = 0; i < this.length; i++) {
+        newArray[i] = predicate(this[i], i, this);
```
Comment (Shalimov) [inline answers.js]: good,
but I'd prefer to use `.push` less deps from outer indexes
```example
array.push(val)
```

File: answers.js [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    return newArray;
+};
+
+Array.prototype.each = function(predicate) {
+    if (typeof predicate !== "function") {
+        console.log("predicate must be a function");
```
Comment (Shalimov) [inline answers.js]: irritates me

File: answers.js [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+    module.take = function(end) {
+        return this.slice(0, end);
+    };
+
+    function paramCheckDecorator(type, fn) {
+        return function(param) {
+            if (typeof param !== type) {
+                throw new Error('Predicate should be a function');
+                return this;
+            }
```
Comment (Shalimov) [inline answers.js]: Bullshit, cuz it should do only one thing:
throw an error or return result 
pls choose ur destiny) ahaha

File: answers.js [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+
+            newArray.push(this[i]);
+        }
+
+        return newArray;
+    });
+
+    module.filter = paramCheckDecorator('function', function(predicate) {
+
```
Comment (Shalimov) [inline answers.js]: Get rid of useless empty line

File: answers.js [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+        let newArray = [];
+        for (let i = 0; i < this.length; i++) {
```
Comment (Shalimov) [inline answers.js]: Add a line to separate it in order to make it more readable

Comment (Shalimov) [inline answers.js]: relates to code below too

File: answers.js [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+
+    module.myFind = paramCheckDecorator('function', function(predicate) {
+
+        for (let i = 0; i < this.length; i++) {
+            if (predicate(this[i], i, this) === true) {
+                return this[i];
+            }
+        }
+
+        return udefined;
```
Comment (Shalimov) [inline answers.js]: incorrect shit!!!
this module doesn't work!
Did u check the changes before committing?

File: answers.js [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+
+    module.myFindLast = paramCheckDecorator('function', function(predicate) {
+
+        for (let i = this.length - 1; i >= 0; i--) {
+            if (predicate(this[i], i, this) === true) {
+                return this[i];
+            }
+        }
+
+        return udefined;
```
Comment (Shalimov) [inline answers.js]: There is no such a word!

File: answers.js [hunk 19]
```diff
@@ -0,0 +1,220 @@
+(function(module) {
+    module.first = function() {
```
Comment (Shalimov) [inline answers.js]: add 'use strict' at the top of module declaration
of course inside this IIEF

File: answers.js [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+        return function() {
+            /*   if (typeof this !== 'object' && (typeof obj !== 'function' || obj === null)) {
+                   throw new TypeError('Object.keys called on non-object');
+                   return this;
+               }*/
+            return fn.apply(this, arguments);
+        };
+    }
+
+    module.myKeys = paramCheckDecorator(function() {
```
Comment (Shalimov) [inline answers.js]: What for u use decorator here?

File: answers.js [hunk 21]
```diff
@@ -0,0 +1,220 @@
...
+        }
+
+        return -1;
+    });
+
+})(Array.prototype);
+
+
+(function(module) {
+    function paramCheckDecorator(fn) {
```
Comment (Shalimov) [inline answers.js]: What for?

File: answers.js [hunk 22]
```diff
@@ -0,0 +1,220 @@
...
+            return fn.apply(this, arguments);
+        };
+    }
+
+    module.myKeys = paramCheckDecorator(function() {
+        let hasOwnProperty = Object.prototype.hasOwnProperty;
+        let result = [],
+            prop;
+        for (prop in this) {
+            if (hasOwnProperty.call(this, prop)) {
```
Comment (Shalimov) [inline answers.js]: Good point) Yep
But it seems u steal it from somewhere
(Could u explain me why do u need to use this function outside of object)

File: answers2.js [hunk 23]
```diff
@@ -0,0 +1,63 @@
+/**
+ * Created by anton on 06.12.2016.
+ */
+var Libra = {};
+(function(module) {
+    module.myAssign = function(target) {
```
Comment (Shalimov) [inline answers2.js]: what for u use `my` prefix?
is there any other function `assign` in YOUR library?

File: answers2.js [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+            }
+            let keysArray = Object.keys(Object(nextSource));
+            for (let j = 0; j < keysArray.length; j++) {
+                newObject[keysArray[j]] = nextSource[keysArray[j]]
+            }
+        }
+        return newObject;
+    };
+
+    module.myInherits = function(target) {
```
Comment (Shalimov) [inline answers2.js]: Hey hey (`my`) 
wasn't too obvious but I guessed eventually

Comment (Shalimov) [inline answers2.js]: Ophs what the heck?
assign again?

File: answers.js [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+        }
+        return result;
+    };
+
+    module.myValues = function() {
+        let hasOwnProperty = Object.prototype.hasOwnProperty;
+        let result = [],
+            prop;
+        for (prop in this) {
+            if (hasOwnProperty.call(this, prop)) {
```
Comment (Shalimov) [inline answers.js]: What for u did that?
What I mean is assignment of separated variable to keep hasOwnProperty method

Comment (Shalimov) [inline answers.js]: The same question for all methods from the module!

File: answers.js [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+                let tmp = [];
+                if (hasOwnProperty.call(obj, prop)) {
+                    tmp.push(prop);
+                    tmp.push(obj[prop]);
+                    result.push(tmp);
```
Comment (Shalimov) [inline answers.js]: hmm, it could be more compact in case
```
result.push([ prop, obj[prop] ]) 
```

File: answers.js [hunk 27]
```diff
@@ -0,0 +1,299 @@
...
+})(Object.prototype);
+
+
+(function(module) {
+    'use strict'
+
+    function paramCheckDecorator(fn) {
+        return function(obj) {
+            if (obj !== undefined && typeof obj !== 'object' && (typeof obj !== 'function' || obj === null)) {
+                throw new TypeError('Object.keys called on non-object');
```
Comment (Shalimov) [inline answers.js]: Predefined message that relates to `.keys` method. 
Is it right?

File: answers.js [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+
+})(Object.prototype);
+
+
+(function(module) {
+    'use strict'
+
+    function paramCheckDecorator(fn) {
+        return function(obj) {
+            if (obj !== undefined && typeof obj !== 'object' && (typeof obj !== 'function' || obj === null)) {
```
Comment (Shalimov) [inline answers.js]: Really? What happens if `obj` is equal to `undefined`?

File: answers.js [hunk 29]
```diff
@@ -0,0 +1,299 @@
...
+
+            return fn.apply(this, arguments);
+        }
+    }
+
+    module.myKeys = paramCheckDecorator(function(obj) {
+        let hasOwnProperty = Object.prototype.hasOwnProperty;
+        let result = [],
+            prop;
+        if (obj !== undefined) {
```
Comment (Shalimov) [inline answers.js]: Never mind, just philosophical question, what if decorator can help here?

File: answers.js [hunk 30]
```diff
@@ -0,0 +1,299 @@
...
+            }
+            return result;
+        }
+    });
+
+    module.myValues = paramCheckDecorator(function(obj) {
+        let hasOwnProperty = Object.prototype.hasOwnProperty;
+        let result = [],
+            prop;
+        if (obj !== undefined) {
```
Comment (Shalimov) [inline answers.js]: Pay attention to comment above

File: answers.js [hunk 31]
```diff
@@ -0,0 +1,299 @@
...
+            }
+            return result;
+        }
+    });
+
+    module.myEntries = paramCheckDecorator(function(obj) {
+        let hasOwnProperty = Object.prototype.hasOwnProperty;
+        let result = [],
+            prop;
+        if (obj !== undefined) {
```
Comment (Shalimov) [inline answers.js]: Same shit that above

File: answers2.js [hunk 32]
```diff
@@ -0,0 +1,63 @@
...
+        if (target === undefined || target === null) {
+            throw new TypeError('Cannot convert first argument to object');
+        }
+        let newObject = Object(target);
```
Comment (Shalimov) [inline answers2.js]: Чё за ЩИТ? So close to make me angry

File: answers2.js [hunk 33]
```diff
@@ -0,0 +1,63 @@
+/**
+ * Created by anton on 06.12.2016.
+ */
+var Libra = {};
+(function(module) {
+    module.myAssign = function(target) {
+        if (target === undefined || target === null) {
+            throw new TypeError('Cannot convert first argument to object');
+        }
+        let newObject = Object(target);
```
Comment (Shalimov) [inline answers2.js]: Pfff look below

File: answers2.js [hunk 34]
```diff
@@ -0,0 +1,63 @@
...
+
+    module.myInherits = function(target) {
+        if (target === undefined || target === null) {
+            throw new TypeError('Cannot convert first argument to object');
+        }
+        let newObject = Object(target);
+        let parameters = Array.prototype.slice.call(arguments, 1);
+        for (var i = 0; i < parameters.length; i++) {
+            let nextSource = parameters[i];
+            if (nextSource === undefined || nextSource === null) {
```
Comment (Shalimov) [inline answers2.js]: `nextSource == null` makes both of

File: answers.js (hunk 2 above)
Comment (AntonStaravoitau) [inline answers.js]: clear, in future I will use throwing exceptions

File: answers.js (hunk 3 above)
Comment (AntonStaravoitau) [inline answers.js]: clear

File: answers.js (hunk 25 above)
Comment (AntonStaravoitau) [inline answers.js]: I copyed this style in other place, I can't understand difference between call this function inside or outside this object. In this situation

File: answers.js (hunk 26 above)
Comment (AntonStaravoitau) [inline answers.js]: Clear

File: answers2.js (hunk 23 above)
Comment (AntonStaravoitau) [inline answers2.js]: I just afraid of reserved words, but you are right

File: answers2.js (hunk 34 above)
Comment (AntonStaravoitau) [inline answers2.js]: clear

File: answers2.js (hunk 24 above)
Comment (AntonStaravoitau) [inline answers2.js]: This is wrong commit, I use function assign like template. Function inherits has not realized yet

File: answers.js (hunk 27 above)
Comment (AntonStaravoitau) [inline answers.js]: will be returned undefined. I check how should I do, and fix it

File: answers.js (hunk 28 above)
Comment (AntonStaravoitau) [inline answers.js]: typeError is more suitable

**Pull request merged**